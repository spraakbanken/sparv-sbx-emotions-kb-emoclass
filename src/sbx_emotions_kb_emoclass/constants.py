"""Constants used in this project."""

PROJECT_NAME: str = "sbx_sentence_emotional_classification_kb_emoclass"
