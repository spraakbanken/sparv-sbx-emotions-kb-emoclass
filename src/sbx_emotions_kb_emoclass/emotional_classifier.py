"""Emotional classifier."""

import torch
from setfit import SetFitModel  # type: ignore[import-untyped]
from sparv import api as sparv_api

MODEL_NAME = "KBLab/emotional-classification"
MODEL_REVISION = "73f1663770e79ff5c1aa12a38063a13537a02ce0"

LABELS_SWE = {
    0: "KÄNSLA SAKNAS",
    1: "GLÄDJE",
    2: "KÄRLEK/EMPATI",
    3: "ORO/RÄDSLA",
    4: "SORG/BESVIKELSE",
    5: "ILSKA/HAT",
    6: "HOPP/FÖRVÄNTAN",
}

LABELS_ENG = {
    0: "ABSENCE OF EMOTION",
    1: "HAPPINESS",
    2: "LOVE/EMPATHY",
    3: "FEAR/ANXIETY",
    4: "SADNESS/DISAPPOINTMENT",
    5: "ANGER/HATE",
    6: "HOPE/ANTICIPATION",
}

DEFAULT_LANG: str = "en"


logger = sparv_api.get_logger(__name__)


class EmotionalClassifier:
    """Emotional classifier."""

    def __init__(self, model: SetFitModel | None = None, lang: str | None = None) -> None:
        """Emotional classifier.

        Args:
            model: An optional model to use. If no model is given,
                the default model will be loaded.
            lang: The language to generate labels in. (Defaults to 'en').
        """
        if not lang:
            lang = DEFAULT_LANG
        if lang == "en":
            self.labels = LABELS_ENG
        elif lang == "sv":
            self.labels = LABELS_SWE
        else:
            raise ValueError(f"supported langs are 'en' and 'sv'. Given '{lang}'")
        self.model = model or self._default_model()
        if torch.cuda.is_available():
            device = torch.device("cuda")
        else:
            device = torch.device("cpu")
            logger.warning("Using CPU, is GPU available?")
        logger.info("Using device=%s", device)
        self.model = self.model.to(device)

    @classmethod
    def _default_model(cls) -> SetFitModel:
        return SetFitModel.from_pretrained(MODEL_NAME, revision=MODEL_REVISION)

    def predict_emotion_sentence(self, text: str) -> str:
        """Predict emotion on sentence.

        Args:
            text: the text to analyze

        Returns:
            the predicted emotion
        """
        preds = self.model.predict([text], as_numpy=True)

        return self.labels[preds[0]]  # type: ignore
