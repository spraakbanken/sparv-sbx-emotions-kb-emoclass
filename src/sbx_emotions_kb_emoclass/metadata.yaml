# yaml-language-server: $schema=https://github.com/spraakbanken/metadata/blob/main/schema/metadata.json
id: sbx-swe-emotion-transformers-kblab_emotional_classification-parent
abstract: true
task: emotional analysis
plugin_url: https://pypi.org/project/sparv-sbx-emotions-kb-emoclass
types: analysis
language_codes:
  - swe
keywords:
  - transformers
  - kb-lab
tools:
  - name: "transformers"
    url: "https://github.com/huggingface/transformers"
    license: "Apache-2.0 license"
models:
  - name: "KBLab/emotional-classification"
    url: "https://huggingface.co/KBLab/emotional-classification"
    description: |-
      This is a [SetFit](https://github.com/huggingface/setfit) model that can
      be used for classification of emotions in Swedish text. The model supports
      seven basic emotions.
    license: "Apache-2.0 license"
other_references:
  - "Repository: https://github.com/spraakbanken/sparv-sbx-emotions-kb-emoclass"
created: 2025-10-17
updated: 2025-10-24
---
id: sbx-swe-emotion_sentence-transformers-kblab_emotional_classification
parent: sbx-swe-emotion-transformers-kblab_emotional_classification-parent
short_description:
  swe: Känsloanalys på meningsnivå mha KBLab/emotional-classification
  eng: Emotional analysis at sentence level with KBLab/emotional-classification
description:
  swe: |-
    Känsloanalys på meningsnivå med hjälp av [transformers](https://github.com/Huggingface/transformers)
    och [KBLab/emotional-classification](https://huggingface.co/KBLab/emotional-classification).
  eng: |-
    Emotional analysis at sentence level using [transformers](https://github.com/Huggingface/transformers)
    and [KBLab/emotional-classification](https://huggingface.co/KBLab/emotional-classification).
analysis_unit: sentence
annotations:
  - <sentence>:sbx_emotions_kb_emoclass.emotion--kb-emoclass
example_output: |-
  ```xml
  <sentence emotion--kb-emoclass="SADNESS/DISAPPOINTMENT">
  <token pos="PN">Jag</token>
  <token pos="VB">är</token>
  <token pos="PC">förvånad</token>
  <token pos="MID">,</token>
  <token pos="PC">chockad</token>
  <token pos="KN">och</token>
  <token pos="PC">bestört</token>
  <token pos="PP">över</token>
  <token pos="PN">det</token>
  <token pos="AB">här</token>
  <token pos="MID">,</token>
  <token pos="VB">säger</token>
  <token pos="PM">Anders</token>
  <token pos="PM">Persson</token>
  <token pos="PP">till</token>
  <token pos="PM">SVT</token>
  <token pos="PM">Nyheter</token>
  <token pos="PM">Småland</token>
  <token pos="MAD">.</token>
  </sentence>
  ```
