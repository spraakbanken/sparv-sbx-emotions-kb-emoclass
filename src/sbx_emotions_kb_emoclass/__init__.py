"""Sparv plugin for annotating sentences with emotional classification."""

from sparv import api as sparv_api

from sbx_emotions_kb_emoclass.annotations import annotate_sentence
from sbx_emotions_kb_emoclass.constants import PROJECT_NAME

__all__ = ["annotate_sentence"]


__description__ = "Annotate sentences with emotional classification."
__version__ = "0.1.1"

__config__ = [
    sparv_api.Config(
        f"{PROJECT_NAME}.annotation_lang",
        description="The language to use in the annotations, currently 'en' or 'sv'.",
        default=None,
    )
]
