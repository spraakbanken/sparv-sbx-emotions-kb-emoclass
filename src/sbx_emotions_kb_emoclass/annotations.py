"""Annotations for emotional classification."""

from sparv import api as sparv_api  # type: ignore [import-untyped]

from sbx_emotions_kb_emoclass.constants import PROJECT_NAME

logger = sparv_api.get_logger(__name__)

SEP_TOKEN: str = " "


@sparv_api.annotator("Emotional classification of sentences", language=["swe"])
def annotate_sentence(
    out_sentence_emotions: sparv_api.Output = sparv_api.Output(
        f"<sentence>:{PROJECT_NAME}.emotion--kb-emoclass",
        description="Emotional analysis of sentence with KBLab/emotional-classification",
    ),
    word: sparv_api.Annotation = sparv_api.Annotation("<token:word>"),
    sentence: sparv_api.Annotation = sparv_api.Annotation("<sentence>"),
    annotation_lang: str | None = sparv_api.Config(f"{PROJECT_NAME}.annotation_lang"),
) -> None:
    """Sentiment analysis of sentence with KBLab/robust-swedish-sentiment-multiclass."""
    from sbx_emotions_kb_emoclass.emotional_classifier import (  # noqa: PLC0415
        EmotionalClassifier,
    )

    try:
        classifier = EmotionalClassifier(lang=annotation_lang)
    except ValueError as exc:
        raise sparv_api.SparvErrorMessage(
            f"'{PROJECT_NAME}.annotation_lang' got an unrecognized value: '{annotation_lang}'"
        ) from exc
    sentences, _orphans = sentence.get_children(word)
    token_word = list(word.read())
    out_sentence_emotions_annotation = sentence.create_empty_attribute()

    logger.progress(total=len(sentences))  # type: ignore
    for sent_i, sent in enumerate(sentences):
        sent_to_tag = SEP_TOKEN.join([token_word[token_index] for token_index in sent])
        out_sentence_emotions_annotation[sent_i] = classifier.predict_emotion_sentence(
            sent_to_tag
        )
        logger.debug("annotation[%d]=%s", sent_i, out_sentence_emotions_annotation[sent_i])

    out_sentence_emotions.write(out_sentence_emotions_annotation)
