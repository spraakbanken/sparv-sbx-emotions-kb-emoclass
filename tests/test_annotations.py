import pytest
from sparv import api as sparv_api
from sparv_testing import MemoryOutput, MockAnnotation
from syrupy.assertion import SnapshotAssertion

from sbx_emotions_kb_emoclass.annotations import annotate_sentence


def test_annotate_sentence_lang_en(snapshot) -> None:  # noqa: ANN001
    output: MemoryOutput = MemoryOutput()
    word = MockAnnotation(
        name="<token>",
        values=[
            "Han",
            "var",
            "glad",
            ".",
            "Rihanna",
            "uppges",
            "gravid",
            ".",
            "Jag",
            "har",
            "채tit",
            "s채mre",
            ".",
        ],
    )
    sentence = MockAnnotation(
        name="<sentence>", children={"<token>": [[0, 1, 2, 3], [4, 5, 6, 7], [8, 9, 10, 11, 12]]}
    )

    annotate_sentence(output, word, sentence, None)

    assert output.values == snapshot


def test_annotate_sentence_lang_sv(snapshot) -> None:  # noqa: ANN001
    output: MemoryOutput = MemoryOutput()
    word = MockAnnotation(
        name="<token>",
        values=[
            "Han",
            "var",
            "glad",
            ".",
            "Rihanna",
            "uppges",
            "gravid",
            ".",
            "Jag",
            "har",
            "채tit",
            "s채mre",
            ".",
        ],
    )
    sentence = MockAnnotation(
        name="<sentence>", children={"<token>": [[0, 1, 2, 3], [4, 5, 6, 7], [8, 9, 10, 11, 12]]}
    )

    annotate_sentence(output, word, sentence, "sv")

    assert output.values == snapshot


def test_annotate_sentence_with_unsupported_lang_raise_value_error(
    snapshot: SnapshotAssertion,
) -> None:
    with pytest.raises(sparv_api.SparvErrorMessage) as exc:
        annotate_sentence(None, None, "sentence", "xx")

    assert str(exc) == snapshot
