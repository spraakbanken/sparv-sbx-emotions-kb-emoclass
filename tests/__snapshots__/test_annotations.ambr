# serializer version: 1
# name: test_annotate_sentence_lang_en
  list([
    'HAPPINESS',
    'ABSENCE OF EMOTION',
    'SADNESS/DISAPPOINTMENT',
  ])
# ---
# name: test_annotate_sentence_lang_sv
  list([
    'GLÄDJE',
    'KÄNSLA SAKNAS',
    'SORG/BESVIKELSE',
  ])
# ---
# name: test_annotate_sentence_with_unsupported_lang_raise_value_error
  '<ExceptionInfo SparvErrorMessage("<<<START>>>\\n\\n\'sbx_emotions_kb_emoclass.annotation_lang\' got an unrecognized value: \'xx\'<<<END>>>") tblen=2>'
# ---
